File: kotlin/advent/src/main/kotlin/org/gristle/pdxpuzzles/advent/y2015/Y15D03.kt
package org.gristle.pdxpuzzles.advent.y2015

import org.gristle.pdxpuzzles.advent.utilities.Day
import org.gristle.pdxpuzzles.advent.utilities.iteration.collate
import org.gristle.pdxpuzzles.utilities.objects.Coord

class Y15D03(private val input: String) : Day {

    private fun deliver(s: String) = s
        .runningFold(Coord.ORIGIN) { santa, dir -> santa.move(dir) }
        .toSet()

    override fun part1() = deliver(input).size

    override fun part2(): Int = input
        .collate(2)
        .map(::deliver)
        .reduce(Set<Coord>::union)
        .size
}

fun main() = Day.runDay(Y15D03::class)

//    Class creation: 21ms
//    Part 1: 2081 (8ms)
//    Part 2: 2341 (6ms)
//    Total time: 36ms

File: rust/advent/src/bin/2015/y15d03.rs
use advent::utilities::get_input::get_input;
use rustc_hash::FxHashSet;
use utilities::{collation::Collate, structs::{coord::Coord2, stopwatch::{ReportDuration, Stopwatch}}};

type Input<'a> = &'a str;
type Output = usize;
type Pos = Coord2;

fn main() {
    let mut stopwatch = Stopwatch::new();
    stopwatch.start();
    let input = get_input(15, 3).unwrap();
    println!("Input parsed ({})", stopwatch.lap().report());
    println!("1. {} ({})", part1(&input), stopwatch.lap().report());
    println!("2. {} ({})", part2(&input), stopwatch.lap().report());
    println!("Total: {}", stopwatch.stop().report());
}

fn deliver(chars: impl Iterator<Item = char>) -> FxHashSet<Pos> {
    std::iter::once(Pos::origin())
        .chain(chars.scan(Pos::origin(), |state, c| {
            *state = state.char_move(c).unwrap();
            Some(*state)
        }))
        .collect()
}

fn part1(input: Input) -> Output {
    deliver(input.chars()).len()
}

fn part2(input: Input) -> Output {
    input.chars()
        .collate(2)
        .into_iter()
        .map(|thread| deliver(thread.into_iter()))
        .reduce(|acc, other| {
            acc.union(&other).copied().collect()
        })
        .unwrap()
        .len()
}

#[test]
fn default() {
    let input = get_input(15, 3).unwrap();
    assert_eq!(2081, part1(&input));
    assert_eq!(2341, part2(&input));
}

// Input parsed (224μs)
// 1. 2081 (345μs)
// 2. 2341 (360μs)
// Total: 933μs
